File: go/src/v.io/x/browser/sample/mocks/speaker.go
// Copyright 2015 The Vanadium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package mocks

import (
	"errors"
	"fmt"

	"v.io/v23/rpc"
	"v.io/x/browser/sample"
)

const speakerDefaultVolume = uint16(10)

// Speaker allows clients to control the music being played.
type speaker struct {
	currentSong    string
	playing        bool
	volume         uint16
	speakerLibrary map[string]bool
}

// Play starts or continues the current song.
func (s *speaker) Play(rpc.ServerCall) error {
	if s.currentSong == "" {
		return errors.New("no current song")
	}
	s.playing = true
	return nil
}

// PlaySong plays back the given song title, if possible.
func (s *speaker) PlaySong(_ rpc.ServerCall, title string) error {
	if !s.speakerLibrary[title] {
		return errors.New(fmt.Sprintf("%q does not exist", title))
	}
	s.currentSong = title
	s.playing = true
	return nil
}

// PlayStream plays the given stream of music data.
func (s *speaker) PlayStream(sample.SpeakerPlayStreamServerCall) error {
	s.currentSong = ""
	s.playing = true
	return nil
}

// GetSong retrieves the title of the Speaker's current song, if any.
func (s *speaker) GetSong(rpc.ServerCall) (string, error) {
	return s.currentSong, nil
}

// Pause playback of the Speaker's current song.
func (s *speaker) Pause(rpc.ServerCall) error {
	s.playing = false
	return nil
}

// Stop playback of the Speaker's current song.
func (s *speaker) Stop(rpc.ServerCall) error {
	s.currentSong = ""
	s.playing = false
	return nil
}

// Volume adjusts the Speaker's volume.
func (s *speaker) Volume(_ rpc.ServerCall, volume uint16) error {
	s.volume = volume
	return nil
}

// GetVolume retrieves the Speaker's volume.
func (s *speaker) GetVolume(rpc.ServerCall) (uint16, error) {
	return s.volume, nil
}

// AddSongs adds the list of given songs to the song library.
func (s *speaker) AddSongs(_ rpc.ServerCall, songs []string) error {
	for _, song := range songs {
		s.speakerLibrary[song] = true // No-op if the song is there.
	}
	return nil
}

// Delete removes the list of given songs from the song library.
func (s *speaker) Delete(_ rpc.ServerCall, songs []string) error {
	for _, song := range songs {
		delete(s.speakerLibrary, song) // No-op if the song isn't there.
		if s.currentSong == song {     // Stop playing the current song if it was removed.
			s.currentSong = ""
			s.playing = false
		}
	}
	return nil
}

// NewSpeaker creates a new speaker stub.
func NewSpeaker() *speaker {
	return &speaker{
		playing: false,
		volume:  speakerDefaultVolume,
		speakerLibrary: map[string]bool{ // Start with some default songs.
			"Happy Birthday":          true,
			"Never Gonna Give You Up": true,
		},
	}
}

File: src/services/sample-world/speaker.js
// Copyright 2015 The Vanadium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

var vdl = require('./ifc');

module.exports = Speaker;

var DEFAULT_VOLUME = 10;
var DEFAULT_SONGS = new Set(['Happy Birthday', 'Never Gonna Give You Up']);

// Speaker allows clients to control the music being played.
function Speaker() {
  this.playing = false;
  this.volume = DEFAULT_VOLUME;
  this.speakerLibrary = DEFAULT_SONGS;
  this.currentSong = null;
}

Speaker.prototype = new vdl.Speaker();

// Play starts or continues the current song.
Speaker.prototype.play = function(context, serverCall) {
  if (this.currentSong === null) {
    throw new Error('no current song');
  }
  this.playing = true;
};

// PlaySong plays back the given song title, if possible.
Speaker.prototype.playSong = function(context, serverCall, title) {
  if (!this.speakerLibrary.has(title)) {
    throw new Error(title + ' does not exist');
  }
  this.currentSong = title;
  this.playing = true;
};

// PlayStream plays the given stream of music data.
Speaker.prototype.playStream = function(context, serverCall) {
  // Note: Not truly implemented, since this should open a stream.
  this.currentSong = null;
  this.playing = true;
};

// GetSong retrieves the title of the Speaker's current song, if any.
Speaker.prototype.getSong = function(context, serverCall) {
  return this.currentSong;
};

// Pause playback of the Speaker's current song.
Speaker.prototype.pause = function(context, serverCall) {
  this.playing = false;
};

// Stop playback of the Speaker's current song.
Speaker.prototype.stop = function(context, serverCall) {
  this.currentSong = null;
  this.playing = false;
};

// Volume adjusts the Speaker's volume.
Speaker.prototype.volume = function(context, serverCall, volume) {
  this.volume = volume;
};

// GetVolume retrieves the Speaker's volume.
Speaker.prototype.getVolume = function(context, serverCall) {
  return this.volume;
};

// AddSongs adds the list of given songs to the song library.
Speaker.prototype.addSongs = function(context, serverCall, songs) {
  songs.forEach(function(song) {
    this.speakerLibrary.add(song); // No-op if the song is already there.
  });
};

// Delete removes the list of given songs from the song library.
Speaker.prototype.delete = function(context, serverCall, songs) {
  songs.forEach(function(song) {
    this.speakerLibrary.delete(song);
    if (this.currentSong === song) {
      this.currentSong = null;
      this.playing = false;
    }
  });
};